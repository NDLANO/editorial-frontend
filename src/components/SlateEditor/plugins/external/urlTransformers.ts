/**
 * Copyright (c) 2025-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import queryString from "query-string";
import { fetchNrkMedia } from "../../../../modules/video/nrkApi";
import config from "../../../../config";

interface UrlTransformer {
  domains: string[];
  shouldTransform: (url: string, domains: string[]) => boolean;
  transform: (url: string) => Promise<string>;
}

// Fetches media-id from url and returns embed url
const nrkTransformer: UrlTransformer = {
  domains: ["nrk.no", "www.nrk.no"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }

    const oldMediaId = queryString.parse(urlObj.search).mediaId;
    const newMediaId = Number(urlObj.pathname.split("/skole-deling/")[1]);
    const mediaId = newMediaId ? newMediaId : oldMediaId;
    if (mediaId) {
      return true;
    }
    return false;
  },
  transform: async (url) => {
    const urlObj = new URL(url);
    const oldMediaId = queryString.parse(urlObj.search).mediaId;
    const newMediaId = Number(urlObj.pathname.split("/skole-deling/")[1]);
    const mediaId = newMediaId ? newMediaId : oldMediaId;
    if (!mediaId) {
      return url;
    }
    try {
      const nrkMedia = await fetchNrkMedia(mediaId);
      if (nrkMedia.psId) {
        return `https://static.nrk.no/ludo/latest/video-embed.html#id=${nrkMedia.psId}`;
      }
      return url;
    } catch {
      return url;
    }
  },
};

// Replaces www-hostname with embed-hostname
const tedTransformer: UrlTransformer = {
  domains: ["www.ted.com"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.host = obj.host.replace(/www/, "embed");
    return obj.href;
  },
};

// Replaces 'pen' with 'embed' in path
const codepenTransformer: UrlTransformer = {
  domains: ["codepen.io"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);

    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (!urlObj.href.includes("/pen/")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.pathname = obj.pathname.replace(/pen/, "embed");
    const penID = obj.pathname.split("/").pop();
    if (penID) {
      return obj.href;
    }
    return url;
  },
};

// Ensures url ends with /embed
const flourishTransformer: UrlTransformer = {
  domains: ["public.flourish.studio", "flo.uri.sh"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (urlObj.href.endsWith("/embed")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.pathname = obj.pathname.replace(/\/+$/, "") + "/embed";
    return obj.href;
  },
};

// Replaces 'model' with 'embed' in path
const sketchupTransformer: UrlTransformer = {
  domains: ["3dwarehouse.sketchup.com"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (!urlObj.href.includes("/model/")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.pathname = obj.pathname.replace(/model/, "embed");
    const parts = obj.pathname.split("/");
    const index =
      parts.findIndex((part) => part.match(/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/)) ??
      parts.length;
    obj.pathname = parts.slice(0, index + 1).join("/");
    return obj.href;
  },
};

const sketcfabTransformer: UrlTransformer = {
  domains: ["sketchfab.com"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (!urlObj.href.includes("/3d-models/")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const urlObj = new URL(url);
    const embedId = urlObj.href.split("-").pop();
    if (embedId?.match(/\b[0-9a-f]{32}/)) {
      urlObj.pathname = `/models/${embedId}/embed`;
    }
    return urlObj.href;
  },
};

const jeopardyLabTransformer: UrlTransformer = {
  domains: ["jeopardylabs.com"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (!urlObj.href.includes("/play/")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    if (url.endsWith("?embed=1")) {
      return url;
    }
    return url.concat("?embed=1");
  },
};

const gapminderTransformer: UrlTransformer = {
  domains: ["www.gapminder.org"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (urlObj.href.includes("?embedded=true")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const urlObj = new URL(url);
    urlObj.href = urlObj.href.replace("tools/", "tools/?embedded=true");
    return urlObj.href;
  },
};

const norgesfilmTransformer: UrlTransformer = {
  domains: ["ndla.filmiundervisning.no", "ndla2.filmiundervisning.no"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);
    if (!config.norgesfilmNewUrl) {
      return false;
    }
    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (!urlObj.href.includes("ndlafilm.aspx?filmId=")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const urlObj = new URL(url);
    urlObj.href = urlObj.href.replace("ndlafilm.aspx?filmId=", "");
    return urlObj.href;
  },
};

const kartiskolenTransformer: UrlTransformer = {
  domains: ["kartiskolen.no"],
  shouldTransform: (url, domains) => {
    const urlObj = new URL(url);

    if (!domains.includes(urlObj.hostname)) {
      return false;
    }
    if (urlObj.pathname.startsWith("/embed.html")) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const urlObj = new URL(url);
    urlObj.pathname = urlObj.pathname.replace(/\/?$/, "/embed.html");
    return urlObj.href;
  },
};

export const urlTransformers: UrlTransformer[] = [
  nrkTransformer,
  codepenTransformer,
  tedTransformer,
  flourishTransformer,
  sketchupTransformer,
  sketcfabTransformer,
  jeopardyLabTransformer,
  gapminderTransformer,
  norgesfilmTransformer,
  kartiskolenTransformer,
];
