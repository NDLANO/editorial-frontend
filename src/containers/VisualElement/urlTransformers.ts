import queryString from 'query-string';
import { fetchNrkMedia } from '../../modules/video/nrkApi';
import { urlAsATag } from '../../util/htmlHelpers';

export interface UrlTransformer {
  domains: string[];
  shouldTransform: (url: string, domains: string[]) => boolean;
  transform: (url: string) => Promise<string>;
}

// Fetches media-id from url and returns embed url
const nrkTransformer: UrlTransformer = {
  domains: ['nrk.no', 'www.nrk.no'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }

    const oldMediaId = queryString.parse(aTag.search).mediaId;
    const newMediaId = Number(aTag.pathname.split('/skole-deling/')[1]);
    const mediaId = newMediaId ? newMediaId : oldMediaId;
    if (mediaId) {
      return true;
    }
    return false;
  },
  transform: async (url) => {
    const aTag = urlAsATag(url);
    const oldMediaId = queryString.parse(aTag.search).mediaId;
    const newMediaId = Number(aTag.pathname.split('/skole-deling/')[1]);
    const mediaId = newMediaId ? newMediaId : oldMediaId;
    if (!mediaId) {
      return url;
    }
    try {
      const nrkMedia = await fetchNrkMedia(mediaId);
      if (nrkMedia.psId) {
        return `https://static.nrk.no/ludo/latest/video-embed.html#id=${nrkMedia.psId}`;
      }
      return url;
    } catch {
      return url;
    }
  },
};

// Replaces www-hostname with embed-hostname
const tedTransformer: UrlTransformer = {
  domains: ['www.ted.com'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.host = obj.host.replace(/www/, 'embed');
    return obj.href;
  },
};

// Replaces 'pen' with 'embed' in path
const codepenTransformer: UrlTransformer = {
  domains: ['codepen.io'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }
    if (!aTag.href.includes('/pen/')) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.pathname = obj.pathname.replace(/pen/, 'embed');
    const penID = obj.pathname.split('/').pop();
    if (penID) {
      return obj.href;
    }
    return url;
  },
};

// Ensures url ends with /embed
const flourishTransformer: UrlTransformer = {
  domains: ['public.flourish.studio', 'flo.uri.sh'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }
    if (aTag.href.endsWith('/embed')) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    const parts = obj.pathname.split('/').filter((n) => n);
    parts.push('embed');
    obj.pathname = parts.join('/');
    return obj.href;
  },
};

// Replaces 'model' with 'embed' in path
const sketchupTransformer: UrlTransformer = {
  domains: ['3dwarehouse.sketchup.com'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }
    if (!aTag.href.includes('/model/')) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const obj = new URL(url);
    obj.pathname = obj.pathname.replace(/model/, 'embed');
    const parts = obj.pathname.split('/');
    const index =
      parts.findIndex((part) =>
        part.match(/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/),
      ) ?? parts.length;
    obj.pathname = parts.slice(0, index + 1).join('/');
    return obj.href;
  },
};

const sketcfabTransformer: UrlTransformer = {
  domains: ['sketchfab.com'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }
    if (!aTag.href.includes('/3d-models/')) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    const embedId = url.split('-').pop();
    if (embedId?.match(/\b[0-9a-f]{32}/)) {
      return `https://sketchfab.com/models/${embedId}/embed`;
    }
    return url;
  },
};

const jeopardyLabTransformer: UrlTransformer = {
  domains: ['jeopardylabs.com'],
  shouldTransform: (url, domains) => {
    const aTag = urlAsATag(url);

    if (!domains.includes(aTag.hostname)) {
      return false;
    }
    if (!aTag.href.includes('/play/')) {
      return false;
    }
    return true;
  },
  transform: async (url) => {
    if (url.endsWith('?embed=1')) {
      return url;
    }
    return url.concat('?embed=1');
  },
};

export const urlTransformers: UrlTransformer[] = [
  nrkTransformer,
  codepenTransformer,
  tedTransformer,
  flourishTransformer,
  sketchupTransformer,
  sketcfabTransformer,
  jeopardyLabTransformer,
];
