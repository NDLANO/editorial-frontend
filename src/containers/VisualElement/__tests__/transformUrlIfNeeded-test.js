/*
 * Copyright (c) 2019-present, NDLA.
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import nock from 'nock';
import { urlTransformers } from '../urlTransformers';

const transformUrlIfNeeded = async (url) => {
  for (const rule of urlTransformers) {
    if (rule.shouldTransform(url, rule.domains)) {
      return await rule.transform(url);
    }
  }
  return url;
};

test('transformUrlIfNeeded returns static nrk url if correct nrk url is used', async () => {
  nock('http://nrk-api').get('/skole/api/media/23618').reply(200, { psId: '33' });
  const url = await transformUrlIfNeeded('https://www.nrk.no/skole-deling/23618');

  expect(url).toMatchSnapshot();
});

test('transformUrlIfNeeded returns url sent in if nrk api should return undefined', async () => {
  nock('http://nrk-api').get('/skole/api/media/23618').reply(200);
  const url = await transformUrlIfNeeded('https://www.nrk.no/skole-deling/23618');

  expect(url).toMatchSnapshot();
});

test('transformurlifNeeded returns static nrk url for old nrk embed format', async () => {
  nock('http://nrk-api').get('/skole/api/media/23618').reply(200, { psId: '33' });
  const url = await transformUrlIfNeeded('https://www.nrk.no/skole/?mediaId=23618');
  expect(url).toMatchSnapshot();
});

test('transformUrlIfNeeded returns url sent in if nrk api should return something else', async () => {
  nock('http://nrk-api').get('/skole/api/media/23618').reply(200, { somethingElse: 'test' });
  const url = await transformUrlIfNeeded('https://www.nrk.no/skole/?mediaId=23618');

  expect(url).toMatchSnapshot();
});

test('transformUrlIfNeeded returns same url if not from nrk', async () => {
  const url = await transformUrlIfNeeded('https://somerandomurl.com');

  expect(url).toMatchSnapshot();
});

test('transformUrlIfNeeded does not crash without url', async () => {
  const url = await transformUrlIfNeeded();

  expect(url).toMatchSnapshot();
});

test('transformUrlIfNeeded returns codepen embed url', async () => {
  const url1 = await transformUrlIfNeeded('https://codepen.io/user/pen/qCnfB');
  expect(url1).toMatch('https://codepen.io/user/embed/qCnfB');

  const url2 = await transformUrlIfNeeded('https://codepen.io/team/user/pen/rzqYxX');
  expect(url2).toMatch('https://codepen.io/team/user/embed/rzqYxX');
});

test('transformUrlIfNeeded returns original if not pen', async () => {
  const url1 = await transformUrlIfNeeded('https://codepen.io/some-other-url');

  expect(url1).toMatch('https://codepen.io/some-other-url');
});

test('transformUrlIfNeeded replaces www with embed for ted.com', async () => {
  const url1 = await transformUrlIfNeeded('https://www.ted.com/talks/some_fancy_talk');
  expect(url1).toMatch('https://embed.ted.com/talks/some_fancy_talk');
});

test('transformUrlIfNeeded returns original for flourish', async () => {
  const url1 = await transformUrlIfNeeded(
    'https://public.flourish.studio/visualisation/8515737/embed',
  );
  expect(url1).toMatch('https://public.flourish.studio/visualisation/8515737/embed');
  const url2 = await transformUrlIfNeeded('https://flo.uri.sh/visualisation/8515737/embed');
  expect(url2).toMatch('https://flo.uri.sh/visualisation/8515737/embed');
});

test('transformUrlIfNeeded adds /embed for flourish', async () => {
  const url1 = await transformUrlIfNeeded('https://public.flourish.studio/visualisation/8515737/');
  expect(url1).toMatch('https://public.flourish.studio/visualisation/8515737/embed');
  const url2 = await transformUrlIfNeeded('https://flo.uri.sh/visualisation/8515737');
  expect(url2).toMatch('https://flo.uri.sh/visualisation/8515737/embed');
});

test('transformUrlIfNeeded creates embed-url for sketchup model if needed', async () => {
  const url1 = await transformUrlIfNeeded(
    'https://3dwarehouse.sketchup.com/model/eb498ef3-3de5-42ca-a621-34ea29cc08c4/Oppbygging-av-ringmur',
  );
  expect(url1).toMatch(
    'https://3dwarehouse.sketchup.com/embed/eb498ef3-3de5-42ca-a621-34ea29cc08c4',
  );
  const url2 = await transformUrlIfNeeded(
    'https://3dwarehouse.sketchup.com/embed/eb498ef3-3de5-42ca-a621-34ea29cc08c4',
  );
  expect(url2).toMatch(
    'https://3dwarehouse.sketchup.com/embed/eb498ef3-3de5-42ca-a621-34ea29cc08c4',
  );
});
