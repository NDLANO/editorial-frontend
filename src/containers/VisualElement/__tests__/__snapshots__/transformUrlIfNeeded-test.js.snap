// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformUrlIfNeeded does not crash without url 1`] = `undefined`;

exports[`transformUrlIfNeeded returns kahoot embed url if correct kahoot url is used 1`] = `"https://embed.kahoot.it/ab095e8e-1234-1234-1234-a0386f1811b5"`;

exports[`transformUrlIfNeeded returns original if kahoot profile url is used 1`] = `"https://create.kahoot.it/profiles/random/ab095e8e-1234-1234-1234-a0386f1811b5"`;

exports[`transformUrlIfNeeded returns original if wrong kahoot url is used 1`] = `"https://create.kahoot.it/share/random/ab095e8e-1234-1234-1234-a0386f1811b5/test"`;

exports[`transformUrlIfNeeded returns original if wrong kahoot url is used 2`] = `"https://create.kahoot.it/share/random/ab09e8e-1234-1234-1234-a0386f1811b"`;

exports[`transformUrlIfNeeded returns same url if not from nrk 1`] = `"https://somerandomurl.com"`;

exports[`transformUrlIfNeeded returns static nrk url if correct nrk url is used 1`] = `"https://static.nrk.no/ludo/latest/video-embed.html#id=33"`;

exports[`transformUrlIfNeeded returns url sent in if nrk api should return something else 1`] = `"https://www.nrk.no/skole/?mediaId=23618"`;

exports[`transformUrlIfNeeded returns url sent in if nrk api should return undefined 1`] = `"https://www.nrk.no/skole/?mediaId=23618"`;
